### ISSUE TITLE
[TW-1713] Detection of infinite loops caused by hooks

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskwarrior#1737]
_Robin Green on 2015-10-09T19:11:32Z says:_ 

 About once a month, I encounter this issue. I type

task unblocked

and an infinite loop of taskwarrior calling a hook script calling taskwarrior... and so on ad infinitum, results. This eventually manifests itself as a nested taskwarrior instance quitting as it is unable to allocate sufficient amounts of memory.

I guess I could prevent this infinite loop by disabling hooks in taskwarrior invocations in my hook script. However, I think that it would be better if taskwarrior had built-in protection against infinite hook loops itself. For example, there could be a configuration variable maximum_hook_depth that would be decremented every time a nested taskwarrior invocation was detected, and when it reached 0, the nested invocation would stop and error out immediately.

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2015-10-09T19:11:32Z
Modified: 2017-01-16T18:08:16Z
```

**Comment by taskwarrior** [comment]
_Wim Schuermann on 2015-10-24T16:10:56Z says:_ 

 This was discussed when hooks were first implemented. Ultimately the decision on whether a hook script should call Taskwarrior with or without hooks active has to be taken by the hook script developer.

Introducing a "recursion limit", or really just a limit on the number of Taskwarrior instances that can be run at the same time, would introduce a lot of complexity with very little gain. Scripts that failed before would still fail, only earlier - the fact that they need to handle errors doesn't change.

All this said, a (somewhat) clean way of introducing limits, or at least informing hooks of the fact that they are being run from within another hook, would be to introduce an environment variable like $TASKWARRIOR_HOOK_NESTED (name needs work) that behaves as you described. I'd argue that it shouldn't lead to Taskwarrior aborting, but should only be used by hook scripts to decide whether they _really_ ought to add level 1001 to the currently active 1000. It would add a way for hook developers to know when something is wrong without shifting the responsibility for dealing with this to Taskwarrior, where it doesn't belong.

TL;DR: I'm not convinced that anything needs to be done here, but am open for having my mind changed of course. :)

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2017-01-16T18:08:16Z says:_ 

 Closing.  I think we have a good workaround, and a full solution would be very complex.
